packages/kbn-cli-dev-mode/src/base_path_proxy/http2.ts

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Url from 'url';
+import { Agent as HttpsAgent, ServerOptions as TlsOptions } from 'https';
+import http2, { Agent as Http2Agent, AutoRequestOptions } from 'http2-wrapper';
+import http2Proxy from 'http2-proxy';
+import { take } from 'rxjs';
+import { getServerOptions, getServerTLSOptions } from '@kbn/server-http-tools';
+
+import { DevConfig, HttpConfig } from '../config';
+import { Log } from '../log';
+import type { BasePathProxyServer, BasePathProxyServerOptions } from './types';
+import { getRandomBasePath } from './utils';
+
+export class Http2BasePathProxyServer implements BasePathProxyServer {
+  private readonly httpConfig: HttpConfig;
+  private server?: http2.Http2SecureServer;
+  private httpsAgent?: HttpsAgent;
+
+  constructor(
+    private readonly log: Log,
+    httpConfig: HttpConfig,
+    private readonly devConfig: DevConfig
+  ) {
+    this.httpConfig = {
+      ...httpConfig,
+      basePath: httpConfig.basePath ?? `/${getRandomBasePath()}`,
+    };
+  }
+
+  public get basePath() {
+    return this.httpConfig.basePath;
+  }
+
+  public get targetPort() {
+    return this.devConfig.basePathProxyTargetPort;
+  }
+
+  public get host() {
+    return this.httpConfig.host;
+  }
+
+  public get port() {
+    return this.httpConfig.port;
+  }
+
+  public async start(options: BasePathProxyServerOptions) {
+    const serverOptions = getServerOptions(this.httpConfig);
+
+    if (this.httpConfig.ssl.enabled) {
+      const tlsOptions = serverOptions.tls as TlsOptions;
+      this.httpsAgent = new HttpsAgent({
+        ca: tlsOptions.ca,
+        cert: tlsOptions.cert,
+        key: tlsOptions.key,
+        passphrase: tlsOptions.passphrase,
+        rejectUnauthorized: false,
+      });
+    }
+
+    await this.setupServer(options);
+
+    this.log.write(
+      `basepath proxy server running at ${Url.format({
```

Comment:
tiny nit, doesn't need to be fixed

http1 includes the prefix `https?://`

Reply:
Yeah, for the HAPI-based proxy, we retrieve the whole URL from the HAPI server itself, and I couldn't find an equivalent API on node's low-level Server class, so I got lazy...

Anyway, fixed - https://github.com/elastic/kibana/pull/183465/commits/d374a6392bd1e4f7479e80eee6c24026b195652a